### DISCUSSION TITLE
macOS Build Error

### DISCUSSION THREAD

**Discussion by ParityError** [tmrlvi/kaspa-miner#54] [Q&A]
When trying to build from source on macOS:

```
$ cargo build --release -p kaspa-miner -p kaspacuda -p kaspaopencl
error: failed to read `/kaspa-miner/plugins/README.md/Cargo.toml`

Caused by:
  Not a directory (os error 20)
```

[Intel Mac Pro, Dual AMD FirePro]

**Comment by tmrlvi** [comment]
In general:
Cuda is not supported on MacOS
OpenCL is supported, but was never tested. We might need to change some configurations in order to build
It should still be possible to build the base miner 

Specifically
This error seems like wrong invocation of `Cargo`. Can you post the command and the full error?

**Comment by ruslansteiger** [comment]
I have trouble as well to get the miner running on my Mac. When I use following command:

```
cargo install kaspa-miner
```

Following error is shown:

```
error[E0433]: failed to resolve: use of undeclared crate or module `keccak`
 --> /Users/rust/.cargo/registry/src/github.com-1ecc6299db9ec823/kaspa-miner-0.2.0/src/pow/keccak.rs:3:5
  |
3 |     keccak::f1600(state);
  |     ^^^^^^ use of undeclared crate or module `keccak`

For more information about this error, try `rustc --explain E0433`.
error: could not compile `kaspa-miner` due to previous error
error: failed to compile `kaspa-miner v0.2.0`, intermediate artifacts can be found at `/var/folders/n0/_wn0_9s97nq8t91l8h6_7lxw0000gn/T/cargo-install0kMI7m`
```

Any ideas how to fix this?

<details>
<summary>Mac specs</summary>

|               | Specificaton  |
| ------------- | ------------- |
| Model         | MacBook Pro   |
| Chip          | Apple M1 Max  |
| OS            | Ventura 13.3  |

</details>

**Comment by tmrlvi** [comment]
@ruslansteiger Note that the command you run references Elichai's miner (the parent of this fork). I'll answer here anyway since it is also relevant here.

The inline assembly doesn't work for M1. To disable it, run with `-F no-asm`
Note that this will not solve the issue of compiling for GPU usage. If you want to install only for cpu usage, you can run:

```{sh}
cargo install --git https://github.com/tmrlvi/kaspa-miner -F no-asm
```

Though it is probably better to use Elichai's miner directly with 
```{sh}
cargo install kaspa-miner -F no-asm
```
